runtime/src/accounts_db.rs

```diff
@@ -3660,79 +3656,210 @@ impl AccountsDB {
+        let items: Vec<_> = snapshot_storages.iter().flatten().collect();
+
+        // Without chunks, we end up with 1 output vec for each outer snapshot storage.
+        // This results in too many vectors to be efficient.
+        const MAX_ITEMS_PER_CHUNK: usize = 5_000;
+        items
+            .par_chunks(MAX_ITEMS_PER_CHUNK)
+            .map(|storages: &[&Arc<AccountStorageEntry>]| {
                 let mut retval = B::default();
-                accounts.into_iter().for_each(|stored_account| {
-                    scan_func(
-                        LoadedAccount::Stored(stored_account),
-                        &mut retval,
-                        storage.slot(),
-                    )
-                });
+
+                for storage in storages {
+                    let accounts = storage.accounts.accounts(0);
+                    accounts.into_iter().for_each(|stored_account| {
+                        scan_func(
+                            LoadedAccount::Stored(stored_account),
+                            &mut retval,
+                            storage.slot(),
+                        )
+                    });
+                }
                 retval
             })
             .collect()
     }
 
-    fn remove_zero_balance_accounts(
-        account_maps: DashMap<Pubkey, CalculateHashIntermediate>,
-    ) -> Vec<(Pubkey, Hash, u64)> {
-        type ShardType = dashmap::lock::RwLock<
-            std::collections::HashMap<
-                solana_sdk::pubkey::Pubkey,
-                dashmap::SharedValue<CalculateHashIntermediate>,
-            >,
-        >;
-        let shards: &[ShardType] = account_maps.shards();
-
-        let hashes: Vec<_> = shards
-            .par_iter()
-            .map(|x| {
-                let a: dashmap::lock::RwLockReadGuard<HashMap<_, _>> = x.read();
-                let res: Vec<_> = a
-                    .iter()
-                    .filter_map(|inp| {
-                        let (pubkey, sv) = inp;
-                        let item = sv.get();
-                        if item.lamports != ZERO_RAW_LAMPORTS_SENTINEL {
-                            Some((*pubkey, item.hash, item.lamports))
-                        } else {
-                            None
-                        }
-                    })
-                    .collect();
-                res
+    fn flatten_hash_intermediate(
+        data_sections_by_pubkey: Vec<Vec<CalculateHashIntermediate>>,
+    ) -> (Vec<CalculateHashIntermediate>, Measure, usize) {
+        let mut flatten_time = Measure::start("flatten");
+        let result: Vec<_> = data_sections_by_pubkey.into_iter().flatten().collect();
+        let raw_len = result.len();
+        flatten_time.stop();
+        (result, flatten_time, raw_len)
+    }
+
+    fn compare_two_hash_entries(
+        a: &CalculateHashIntermediate,
+        b: &CalculateHashIntermediate,
+    ) -> std::cmp::Ordering {
```

Comment:
Can we implement this as `PartialOrd` for `CalculateHashIntermediate` https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html

Reply:
done